cxx-qt-gen/src/cxx_qt.h

```diff
@@ -73,53 +84,100 @@ class CxxQObject : public QObject
   // TODO: we need to document the existence of UpdateStateEvent for users who
   // want to create custom classes that derive from CxxQObject so that they
   // know to avoid clashes with it.
-  static const QEvent::Type UpdateStateEvent;
+  static const QEvent::Type ProcessQueueEvent;
 
 public:
   CxxQObject(QObject* parent = nullptr)
     : QObject(parent)
   {}
   virtual ~CxxQObject() = default;
 
-  void requestUpdate()
+  void requestEmitSignal(std::function<void()> signalFunctor)
   {
-    if (!m_waitingForUpdate.exchange(true, std::memory_order_relaxed))
-      QCoreApplication::postEvent(this, new QEvent(UpdateStateEvent));
+    // Lock the queue, post the event, add to the queue
+    // worst case we'll push an event that does nothing if takeQueue() is
+    // waiting on the lock
+    const std::lock_guard<std::mutex> guard(m_queueMutex);
+
+    if (!m_waitingForUpdate.exchange(true, std::memory_order_relaxed)) {
+      QCoreApplication::postEvent(this, new QEvent(ProcessQueueEvent));
+    }
+
+    m_queue.push_back(std::make_pair(QueueEvent::EmitSignal, signalFunctor));
+  }
+
+  void requestPropertyChange(std::function<void()> propertyFunctor)
+  {
+    // Lock the queue, post the event, add to the queue
+    // worst case we'll push an event that does nothing if takeQueue() is
+    // waiting on the lock
+    const std::lock_guard<std::mutex> guard(m_queueMutex);
+
+    if (!m_waitingForUpdate.exchange(true, std::memory_order_relaxed)) {
+      QCoreApplication::postEvent(this, new QEvent(ProcessQueueEvent));
+    }
+
+    m_queue.push_back(
+      std::make_pair(QueueEvent::UpdatePropertyChange, propertyFunctor));
+  }
+
+  void requestUpdate(std::function<void()> updateFunctor)
+  {
+    // Lock the queue, post the event, add to the queue
+    // worst case we'll push an event that does nothing if takeQueue() is
+    // waiting on the lock
+    const std::lock_guard<std::mutex> guard(m_queueMutex);
+
+    if (!m_waitingForUpdate.exchange(true, std::memory_order_relaxed)) {
+      QCoreApplication::postEvent(this, new QEvent(ProcessQueueEvent));
+    }
+
+    // Compress request updates to only one
+    //
+    // TODO: should we compress events? what happens if we change a
+    // property/emit signal, request update, change property/emit signal,
+    // request update? The second request update won't happen?
+    // Or should the request update always happen after the property/signal
+    // changes?
+    if (std::none_of(
```

Comment:
I think this compression is really problematic with the current function interface.
The updateFunctor could theoretically do whatever, so simply dropping updateRequests, without checking whether the functor does the same thing (which you basically can't) is potentially problematic.

Either the updateFunctor is created within the requestUpdate method, which means it wouldn't need the argument at all, or we consolidate `requestUpdate`, `requestEmitSignal` and `requestPropertyChange` into a single function.
Like I mentioned earlier, this could be called e.g. `runOnGUIThread`.

In this case the m_queue could just be a queue of functions, instead of a pair of events and functions.
Currently either the function isn't really necessary, because the EventType already defines what ought to happen, or the eventtype is not really needed because what happens is entirely defined by the functor, so why have both?

Reply:
If you want to keep the event-information, it would also be possible to use a `std::queue<std::variant<QueueEvent, std::function>>` in which case the action to be done is either defined by the QueueEvent, or by the function.

This would likely be more efficient for the updateRequest case, where the action to be taken is already known and we don't need to go through the std::function. However I'm also fine with just using std::functions for now, as that's more generic.

Reply:
As discussed, removed the compression of events for now (if the user wants to compress events they can do that in their code for now). So now we have `runOnGUIThread` and just `std::queue<std::function<void()>>`